### PR TITLE
initial readme content

### PR DESCRIPTION
**Pull request by linsun**
based on k8s's enhancement repo readme

### CODE REVIEW

**Comment by linsun** [comment]
@jdumars @louiscryan some initial readme PTAL

**Comment by linsun** [comment]
WG leads (and TOC members (@istio/technical-oversight-committee), PTAL and provide feedback.
@duderino @sdake @howardjohn @ostromart @ayj @douglas-reid @mandarjog @esnible @myidpt @liminw @brian-avery @fpesce @ericvn

File: README.md
Comment (howardjohn) [inline README.md]: When should I file an enhancement here and when should I file one as a "type/enhancement" under istio/istio?

File: README.md
Comment (howardjohn) [inline README.md]: nit: Not sure owned by TOC is needed here, at least on the first line

File: README.md
Comment (howardjohn) [inline README.md]: we do not have SIGs

File: README.md
Comment (ayj) [inline README.md]: What's the difference between an issue and enhancement in this context?

File: README.md
Comment (ayj) [inline README.md]:
```suggestion
As a project it is vital to be able to track the chain of custody for a proposed enhancement from conception through implementation. The enhancement repository provides a standardized mechanism for describing important enhancements delivered through Istio releases. This will not only enable the contributors, maintainers and release managers to explaining features to users coherently but also allow adopters to plan their adoption strategy based on a forward looking road map.
```

File: README.md
Comment (howardjohn) [inline README.md]: Does a KEP always imply the creators of the KEP will work on it? Or is it valid to create a full design doc + roadmap but not commit to work on it?

File: README.md
Comment (howardjohn) [inline README.md]: Does IEP replace design docs? I assume the blue ones replaced, maybe not green ones?

File: README.md
Comment (howardjohn) [inline README.md]: Where do I do that? Say someone creates an IEP for feature foobar and I want to discuss it. Do I create a new issue? in what repo?

File: README.md
Comment (howardjohn) [inline README.md]: Why does every enhancement need TOC approval?

File: README.md
Comment (howardjohn) [inline README.md]: Why do we have an RFC document and an enhancement issue?

File: README.md
Comment (howardjohn) [inline README.md]: Why is the release manager involved here? As a release manager, how do I know if I should label an enhancement for my release?

File: README.md
Comment (howardjohn) [inline README.md]: Why is company relevant? Tim's keynote slide "project over company" seems relevant here

File: README.md
Comment (howardjohn) [inline README.md]: Is there not a document checked in as a PR like KEP does it? https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/sidecarcontainers.md

File: README.md
Comment (howardjohn) [reply README.md]: oh wow I misread this. I thought you meant the KEP needed to be approved by more than one company, but you are suggesting each PR associated with that has to be approved by more than one company? That is a very strong claim, and makes Istio feel like less of an open project

File: README.md
Comment (ayj) [reply README.md]: Are enhancement proposals github issues or files checked into the enhancement repo?

File: README.md
Comment (sdake) [inline README.md]: as K2SO would say - this section is "vague and unconvincing" ;-). Defining what an enhancement issue is and how it is created would help. This section simply describes the processes under which an enhancement proposal may be commented on.  Its not entirely clear to the reader that an enhancement issue is what lines 1-54 talk about (tbh, I'm not clear that is what an enhancement issue is).

File: README.md
Comment (sdake) [reply README.md]: for real - this is an onerous requirement...  I think a better approach would be for the TOC to actively monitor enhancements and request a read-out when such a readout is deemed necessary by the TOC.

File: README.md
Comment (sdake) [reply README.md]: I disagree completely with this bullet. The WGs are closer to the work and have more insight into the capacity of the engineering teams and their relative strengths and weaknesses of executing the work.

File: README.md
Comment (sdake) [reply README.md]: I prefer this corporate diversity requirement over no diversity requirement at all. The major key factor in any successful open source project is corporate diversity. The tech can be great, but if it is a single vendor project - it might as well be proprietary.

As for Tim's keynote, I'm not really clear how that is relevant to this community. I have seen many aspirational goals (e.g. OpenStack had project > people > company.  Tim seems to infer that CNCF adopts the model project > company. I think we can all agree we all have the project's best interests in mind when we take actions that affect the sustainability of the project.

File: README.md
Comment (linsun) [reply README.md]: the idea is type/enhancement should be recorded here instead

Comment (linsun) [reply README.md]: issue is a github issue, where enhancement proposal is the proposal linked to the github issue.  I think there are preference that the proposal checked into github after approval.

File: README.md
Comment (linsun) [reply README.md]: I think those are mainly for maintainers, I'd be okay without a strict process as it is not user facing.

File: README.md
Comment (linsun) [reply README.md]: that is totally valid.  I think KEP does require it has engineers allocated to it... We don't necessarily want that... of course, without resource, it will not be placed in any release.

File: README.md
Comment (ayj) [reply README.md]: Commit proposal makes sense to me. The wording here seemed to imply the proposal would only exist as an issue, in which case I didn't see the value over tracking issues in istio/istio.

File: README.md
Comment (ayj) [reply README.md]: Isn't corporate diversity indirectly addressed through the working groups? ToC also has diversity and individual members can /hold if there are concerns that need to be addressed.

File: README.md
Comment (linsun) [reply README.md]: This enhancement repo is mainly for green ones, to make sure everyone understands the scenarios and problems we are solving regardless of the underlying implementation.  The section below on when to comment on is mainly not to clutter an enhancement issue.

File: README.md
Comment (howardjohn) [reply README.md]: Concrete example: https://github.com/istio/istio/issues/19552

Would this be an IEP? I would not think so?

Comment (howardjohn) [reply README.md]: Agree re tech writer, but lets get the content nailed down first then we can improve it

File: README.md
Comment (linsun) [reply README.md]: mark as resolved as it was removed.

File: README.md
Comment (howardjohn) [reply README.md]: Can we decouple this conversation from the rest of the IEP proposal? Lin put this on the TOC agenda which I think is a much more suitable audience. I would suggest we maintain status quo here and keep this PR to the core of adding an IEP process and make changes to our PR process in a separate discussion

File: README.md
Comment (howardjohn) [reply README.md]: I don't see any reason that it needs to be user facing changes only. If I am making a large dev facing change (eg monorepo, hypothetically changing our CI system to CI v2.0) seems reasonable to have an IEP

File: README.md
Comment (douglas-reid) [inline README.md]: Curious: does the "enhancement usually takes multiple releases to complete" imply a definitive feature-flag / opt-in until completion?

And what is implied about features that take only a single release cycle?

File: README.md
Comment (howardjohn) [reply README.md]: I am even more confused then as to what should be an IEP then :slightly_smiling_face:

File: README.md
Comment (douglas-reid) [inline README.md]: Do we have a process for ensuring that all impacted WGs are looped in? Often, impact is not well-understood. Networking changes can have dramatic impact on telemetry without networking teams realizing, etc.

With the fear of creating more process, I might suggest that each IEP *must* contain a direct assessment of impact across all WG-level concerns, and that those sections be reviewed by each impacted WG.

File: README.md
Comment (sdake) [reply README.md]: I suppose we could escalate to the TOC which might escalate to steering related to Istio's lack of corporate diversity.

I (may incorrectly assume) Lin is speaking on behalf of the TOC with this PR. This seems like a really good solution to bake corporate diversity into the project. I assume TOC will approve this README which will set the future direction of enhancements in Istio.

I don't see the value in an early escalation. The TOC can have a level-10 freak out when they review the readme if they find the corporate diversity requirement onerous...

Comment (howardjohn) [reply README.md]: This is not early escalation, its already on the TOC agenda for tomorrow

File: README.md
Comment (sdake) [reply README.md]: As someone that has benefited from an RFC process to create a final decision on a topic that might be contentious, I'd highly encourage (on developer request, but not for every developer-facing change) - that developers be permitted to submit an IEP or whatever we are calling these things.

John lists out several cases where an IEP in my mind as a developer would be absolutely warranted.

A good middle ground is to make this optional for developer-facing changes - and reserve an IEP for a very significant change (consider build tools - all building in containers).  If a community member or maintainer needs the authority of other workgroup leads (or as the document is written the TOC), that should be available to the Istio member as long as it is not used frivolously.

File: README.md
Comment (sdake) [reply README.md]: It seems reasonable, although (as you said) often the impact is not well understood until the implementation is nearing completion. I think we need to trust our working groups to do the right thing, and when they create a crater, to reach out to the impacted working group and determine a solution that is reasonable for both parties.

**Comment by howardjohn** [comment]
Seems like most open source projects are using some variant of https://github.com/kubernetes/enhancements/blob/master/keps/0001-kubernetes-enhancement-proposal-process.md. This includes not just k8s, but python, rust, etc - so its battle tested and most are familiar with it already. Can you explain the diffs and why we are changing them, to help with better understanding what exactly we are proposing and why? Could be a comment here, not necesarily ratified in the readme.md

File: README.md
Comment (douglas-reid) [reply README.md]: My concern, based on experience, is that waiting until near completion to check on impact puts a weird onus on the impacted party to (scramble to) find a way to make something work (as considerable effort has already gone into something and plans/promises have already been made).

If we are going to formalize the planning and roadmap process, I would like us to at least attempt to address the need for fully-articulating impact. We won't always get it right. But, my idealistic hope is that if we, say, have required sections like "Telemetry Collection Impact" and "Environmental Support" in whatever form we adopt, we will catch at least more potential issues than we miss.

Maybe that fits better in the Design Docs ? The relationship between all three is fuzzy to me at the moment.

File: README.md
Comment (ericvn) [inline README.md]: c/explaining/explain/

File: README.md
Comment (ericvn) [reply README.md]: I think what the comment is to not comment on the IEP issue, but maybe in the PR containing the IEP (or maybe as stated in the next section, comments within the RFC (there isn't a PR with a IEP)).

File: README.md
Comment (ericvn) [reply README.md]: I have a similar question. It seems like we are taking part of the KEP process, the creation of the issue, but not the second part with is the actual enhancement request (KEP). It seems we are wanting to continue to use the RFCs. If that's the case, the initial sentence `This repo contains issues and Istio enhancement proposals.` is wrong. It should be written as `This repo contains issues.` The issues would point to RFCs instead of IEPs within the repo.

File: README.md
Comment (ericvn) [inline README.md]: After multiple passes, I don't think this repo will contain Istio Enhancement Proposals (IEPs). It will contain the issue that documents a design doc in Google Docs.

File: README.md
Comment (fpesce) [reply README.md]: It is true that eventually, release managers decide what they will ship or at least delegate this responsibility to the Technical Oversight Committee by challenging it about an alpha/not-tested-enough/dangerously-experimental feature.
It sounds more like an anti-label though: i.e. I-don't-think-I-can-ship-that-in-1.x label.

File: README.md
Comment (fpesce) [reply README.md]: Having 2 companies reviewing each PR might probably exercise a lot of pressure on under-staffed companies / startups. I am not sure this will result into the expected diversity outcome.
Alternatively, maybe it could be required for a PR to be reviewed by at least 1 (one) Working Group leads and/or Maintainer for each Working Group related to that enhancement, as they would have been identified during the `## Is My Thing an Enhancement?` phase.

File: README.md
Comment (mandarjog) [reply README.md]: Is there anything apart from bugs and tests does this leave out ?
If performance improvements are localized, then it does not need interlock with others but it is still a useful as a communication vehicle to show roadmap with some details.

A new testing framework , performance tracking system, would those go here?
Or anything that affects code that *would have* come to TOC comes here ?

File: README.md
Comment (linsun) [reply README.md]: not necessarily, an enhancement will mostly likely finish in multiple releases from code drop + testing + graduation from alpha-beta-stable.  For features take a single release cycle is rare, probably very minor things or updates to an existing feature.  Single release cycle enhancement will be tracked also, when criterias meet.

Comment (linsun) [reply README.md]: Agreed, that is the initial intention.  There is also some preference (cc @jdumars) to document the approved RFC is github (but not the iterations + comments).

File: README.md
Comment (linsun) [reply README.md]: @sdake @howardjohn @brian-avery points taken.  Please suggest changes for it.

We can tag these for developers only as it won't impact release notes.

Comment (linsun) [reply README.md]: @mandarjog I think @sdake mentioned this IEP for developers could be optional which I like.  If the impact is useful for many other contributors to know, it is probably good to have in IEP, e.g. a new testing framework would.

File: README.md
Comment (linsun) [reply README.md]: That is what I was trying to do in the lifecycle portion.

Comment (linsun) [reply README.md]: added a section "Template of an Enhancement Issue" for this.

File: README.md
Comment (linsun) [reply README.md]: appreciate sharing the concerns - I tried to add a section "Template of an Enhancement Issue" for this.

comments is mainly to track requested changes.  issue is unapproved by default, until working group leads approve it.

File: README.md
Comment (linsun) [reply README.md]: Added a section "Template of an Enhancement Issue" for this.

File: README.md
Comment (linsun) [reply README.md]: @ericvn is correct, basically @howardjohn, this is to minimize noises at the enhancement issue and delegate detailed discussion in design doc (blue in google).

**Comment by howardjohn** [comment]
I am a bit confused why are we using issues and approving them rather than PRs which conveniently already have an approval process built in, as well as version control, history, and a more permanent document similar to how other Enhancement Proposals work for other projects?

File: README.md
Comment (linsun) [reply README.md]: folks, understood the concern, I agree projects > company if the project is owned by a foundation.  But istio is not... which is why diversity is critical for istio, especially requirements and user cases.    I reworded the requirements for approval for enhancement issue, which I definitely want corporate diversity.

**Comment by sdake** [comment]
John I have to question this point as well.  GIT delivers on our
requirements and offers data portability.  Reading through the links you
provided on the prior art in this area, there are specific call outs that
issues are unsuitable for tracking such work.

**Comment by linsun** [comment]
@howardjohn I think PR is a solid idea, i like it.

File: README.md
Comment (rshriram) [inline README.md]: nit. `e.g.` not ex :)

Comment (linsun) [reply README.md]: thx!!

**Comment by dgn** [comment]
I really like this. Any chance we can get the ball rolling on this again?

**Comment by linsun** [comment]
Thanks @dgn ! Adding @shwsingh to review this.

**Comment by brian-avery** [comment]
Would love to see this moving again. The feature maturity levels are introducing GitHub templates for the each of the maturity levels. Should enhancement requests have a template as well?

**Review by brian-avery — APPROVED** [review]
Approving this PR. We had a conversation around having a single file for promotion of features in the TOC a couple of weeks ago rather than separate issues for Experimental, Alpha, Beta, and Stable. I wonder if it makes sense to evolve that in lockstep with this such that an enhancement and the requirements that it has met for various feature levels are tracked in the same location. 

I'm not sure if issues are the answer for either. But I'd like to see this merged and start getting feedback on the evolution of enhancements.

Regardless, I think that a good readme describing the process is the first place to start. Thanks for this.

**Comment by linsun** [comment]
Merging this as I believe most comments are addressed.  If you feel **differently**, please open up follow up PRs.

**Pull request merged**
